crowbar_framework/app/models/service_object.rb

```diff
@@ -989,20 +993,28 @@ def apply_role(role, inst, in_queue, bootstrap = false)
       end
     end
 
+    pre_cached_nodes = {}
+    filtered_elements = nil
+
     # When bootstrapping, we don't run chef, so there's no need for queuing
     if bootstrap
-      pre_cached_nodes = {}
       # do not try to process the queue in any case
       in_queue = true
     else
       # Attempt to queue the proposal.  If delay is empty, then run it.
       deps = proposal_dependencies(role)
+      filtered_elements = new_elements.deep_dup
+
+      if skip_unchanged_nodes_enabled
+        filtered_elements = skip_unchanged_nodes(new_elements, old_role, role)
+      end
+
       if skip_unready_nodes_enabled
-        elements_without_unready, pre_cached_nodes = skip_unready_nodes(
-          @bc_name, inst, new_elements, old_elements
+        filtered_elements, pre_cached_nodes = skip_unready_nodes(
+          @bc_name, inst, filtered_elements, old_elements
         )
         delay, pre_cached_nodes = queue_proposal(
-          inst, element_order, elements_without_unready, deps, @bc_name, pre_cached_nodes
+          inst, element_order, filtered_elements, deps, @bc_name, pre_cached_nodes
         )
```

Comment:
This would need to be outside of this `if` statement, since right now if only the first filter is used, we don't enter this code path.

Reply:
yep, indeed, plus with the `cleaned_elements` we can put it outside in any case of the experimental options enabled